carbon-p2-plugin/src/main/java/org/wso2/maven/p2/feature/generate/utils/FeatureFileGeneratorUtils.java

```diff
@@ -0,0 +1,452 @@
+            while ((inputLine = br.readLine()) != null) {
+                stringList.add(inputLine);
+            }
+        } catch (IOException e) {
+            throw new IOException("Error while reading from p2.inf file", e);
+        }
+        return stringList;
+    }
+
+    /**
+     * Generates the feature.xml file for a feature.
+     *
+     * @param resourceBundle containing the project resources
+     * @param featureXmlFile File object representing the feature xml file
+     * @throws TransformerException
+     * @throws IOException
+     * @throws SAXException
+     * @throws ParserConfigurationException
+     */
+    public static void createFeatureXml(FeatureResourceBundle resourceBundle, File featureXmlFile)
+            throws TransformerException, IOException, SAXException, ParserConfigurationException {
+        resourceBundle.getLog().info("Generating feature manifest");
+        Document document = getManifestDocument(resourceBundle.getManifest());
+        Element rootElement = document.getDocumentElement();
+        if (rootElement == null) {
+            rootElement = document.createElement("feature");
+            document.appendChild(rootElement);
+        }
+        if (!rootElement.hasAttribute("id")) {
+            rootElement.setAttribute("id", resourceBundle.getId());
+        }
+        if (!rootElement.hasAttribute("label")) {
+            rootElement.setAttribute("label", resourceBundle.getLabel());
+        }
+        if (!rootElement.hasAttribute("version")) {
+            rootElement.setAttribute("version", BundleUtils.getOSGIVersion(resourceBundle.getVersion()));
+        }
+        if (!rootElement.hasAttribute("provider-name")) {
+            rootElement.setAttribute("provider-name", resourceBundle.getProviderName());
+        }
+
+        NodeList descriptionTags = rootElement.getElementsByTagName("description");
+        if (descriptionTags.getLength() == 0) {
+            Node description = document.createElement("description");
+            description.setTextContent(resourceBundle.getDescription());
+            rootElement.appendChild(description);
+        }
+
+        NodeList copyrightTags = rootElement.getElementsByTagName("copyright");
+
+        if (copyrightTags.getLength() == 0) {
+            Node copyright = document.createElement("copyright");
+            copyright.setTextContent(resourceBundle.getCopyright());
+            rootElement.appendChild(copyright);
+        }
+
+        NodeList licenseTags = rootElement.getElementsByTagName("license");
+        if (licenseTags.getLength() == 0) {
+            Node license = document.createElement("license");
+            ((Element) license).setAttribute("url", resourceBundle.getLicenceUrl());
+            license.setTextContent(resourceBundle.getLicence());
+            rootElement.appendChild(license);
+        }
+
+        ArrayList<Bundle> processedMissingPlugins = getMissingPlugins(resourceBundle.getBundles(), document);
```

Comment:
Would be good to stick to appropriate abstractions for defining variables of this sort in order to maintain better extendability of the implementation.